# INTRODUCTION TO DATA SCIENCE

## Assignment 7

## Topic : Asking and Answering questions to your dataset.

**Dataset which I picked for cleaning of raw data : NYC open data**

** Please use below link for analysing purpose **

- https://data.cityofnewyork.us/Education/SAT-Results/f9bf-2cp4

In [26]:
%matplotlib inline

import os          
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [44]:
fp="C:/Users/kshit/Desktop/IDS assignment 2 review/SAT_Results.csv"
result_data_csv=pd.read_csv(fp)
result_data_csv.head(20) 

DBN                                        SCHOOL NAME  \
0   01M292      HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES   
1   01M448                UNIVERSITY NEIGHBORHOOD HIGH SCHOOL   
2   01M450                         EAST SIDE COMMUNITY SCHOOL   
3   01M458                          FORSYTH SATELLITE ACADEMY   
4   01M509                            MARTA VALLE HIGH SCHOOL   
5   01M515            LOWER EAST SIDE PREPARATORY HIGH SCHOOL   
6   01M539  NEW EXPLORATIONS INTO SCIENCE, TECHNOLOGY AND ...   
7   01M650                               CASCADES HIGH SCHOOL   
8   01M696                     BARD HIGH SCHOOL EARLY COLLEGE   
9   02M047  47 THE AMERICAN SIGN LANGUAGE AND ENGLISH SECO...   
10  02M288                       FOOD AND FINANCE HIGH SCHOOL   
11  02M294                               ESSEX STREET ACADEMY   
12  02M296              HIGH SCHOOL OF HOSPITALITY MANAGEMENT   
13  02M298                                   PACE HIGH SCHOOL   
14  02M300  URBAN ASSEMBLY SCHOOL OF DESIGN AND CONSTRUCTI...   
15  02M303                         FACING HISTORY SCHOOL, THE   
16  02M305  URBAN ASSEMBLY ACADEMY OF GOVERNMENT AND LAW, THE   
17  02M308                       LOWER MANHATTAN ARTS ACADEMY   
18  02M313  JAMES BALDWIN SCHOOL, THE: A SCHOOL FOR EXPEDI...   
19  02M316  URBAN ASSEMBLY SCHOOL OF BUSINESS FOR YOUNG WO...   

   Num of SAT Test Takers SAT Critical Reading Avg. Score SAT Math Avg. Score  \
0                      29                             355                 404   
1                      91                             383                 423   
2                      70                             377                 402   
3                       7                             414                 401   
4                      44                             390                 433   
5                     112                             332                 557   
6                     159                             522                 574   
7                      18                             417                 418   
8                     130                             624                 604   
9                      16                             395                 400   
10                     62                             409                 393   
11                     53                             394                 384   
12                     58                             374                 375   
13                     85                             423                 438   
14                     48                             404                 449   
15                     76                             353                 358   
16                     50                             375                 388   
17                     40                             403                 392   
18                     69                             408                 390   
19                     42                             373                 370   

   SAT Writing Avg. Score  
0                     363  
1                     366  
2                     370  
3                     359  
4                     384  
5                     316  
6                     525  
7                     411  
8                     628  
9                     387  
10                    392  
11                    378  
12                    362  
13                    432  
14                    416  
15                    340  
16                    385  
17                    405  
18                    390  
19                    384

In [45]:
column_names = {'DBN': 'Database_Number', 'SCHOOL NAME': 'School_Name',
                'Num of SAT Test Takers': 'Applicants', 'SAT Critical Reading Avg. Score': 'Avg_CR_score', 
                'SAT Math Avg. Score': 'Avg Math Score', 'SAT Writing Avg. Score': 'Avg Writing Score','Mean':'Mean'} #created a dictionary for column name changes

result_data_csv=(result_data_csv.rename(columns=column_names) 
.dropna(thresh=4)
[['Database_Number','School_Name','Applicants','Avg_CR_score','Avg Math Score','Avg Writing Score']]
.set_index('Database_Number', append=True)
.rename_axis(["SNo", "Database_Number"])
.sort_index())

result_data_csv.head(5) 

School_Name Applicants  \
SNo Database_Number                                                             
0   01M292           HENRY STREET SCHOOL FOR INTERNATIONAL STUDIES         29   
1   01M448                     UNIVERSITY NEIGHBORHOOD HIGH SCHOOL         91   
2   01M450                              EAST SIDE COMMUNITY SCHOOL         70   
3   01M458                               FORSYTH SATELLITE ACADEMY          7   
4   01M509                                 MARTA VALLE HIGH SCHOOL         44   

                    Avg_CR_score Avg Math Score Avg Writing Score  
SNo Database_Number                                                
0   01M292                   355            404               363  
1   01M448                   383            423               366  
2   01M450                   377            402               370  
3   01M458                   414            401               359  
4   01M509                   390            433               384

**Below are the questions which would help us to analyze the data set in a better way.**

## Question 1

## What is the question ?

- Which are the schools having lesser number of SAT takers and the reason behind lesser participants?

## What data will be used to answer the question? 

- To answer the above question i.e What is the reason behind lesser number of participants,we can make use of DBN column from our data set.
- The DBN column helps us to identify the location of the particular school.We can Google search the DBN code and Google provides us with information for the school along with its address.
- For example, DBN 01M458 gives the details of location as '198 Forsyth St, New York, NY 10002'

## Describe a second way to answer the same question?

- An alternate way to provide solution to the above question would be to analyse such schools having lesser number of participants for SAT exam from our dataset.
- We can then analyse the various factors responsible for the anamoly.
- The various factors possible would be lack of proper resources in that region,negligence towards SAT examination etc.

## What are assumptions and limitations of the approach?

**Assumptions for lesser number of SAT takers could be:**
- The primary assumption for this approach is that there are no effective resources allocated to the schools having lesser number of participants.
- The assumption is based upon fact that the Geographical location and facilities of such schools affect a school's ability to allow students to appear for SAT examination.
- We can say,from the dataset,that when a school is located at rural premises,there is a scarce possibility of such a school to have an experienced faculty due to lack or no resources to hire one.
- Also,we see that schools sending higher number of participants for SAT exams are located in developed parts of the state.This helps us to strengthen our approach.

**Limitations for this approach could be:**
- However,there is a limitation of this approach that we cannot for sure analyse the lack of resources and experience for a particular school which send lesser participants for SAT examination.
- One limitation of the proposed approach is that students intending to attend college are more likely to take the SAT and generally have stronger academic credentials than those not taking the exam. This effect is likely to be magnified in states where a low percentage of the student population take the exam, since SAT takers in those states are likely to be high achievers and are less representative of the total student population.
- Also,SAT exam scores could not be the sole criteria for judging a school's prowess in terms of intelligence and quality of education.There could be various other factors which would be critical in determining a school's capability.
- These factors could include difficulty of courses in school,extracurricular activities,letters of recommendation and so on

## What is the statement that you will be able to make after doing the analysis?


- Once the analysis is complete,we can predict the solutions which we would be undertaking to ensure higher educational standards.
- Also, we can try and improve the standards for education in such schools which are located in regions having minimal resources to support success rate for SAT scores.


## ---------------------------------------------------------------------------

## Question 2 

## What is the question? 

- What is the reason for missing or  inconsistent values(s values) in dataset?

In [40]:
list_s = ['s']
result_data_csv[result_data_csv.Avg_CR_score.isin(list_s)].head()

School_Name Applicants  \
SNo Database_Number                                                           
22  02M392                            MANHATTAN BUSINESS ACADEMY          s   
23  02M393                             BUSINESS OF SPORTS SCHOOL          s   
25  02M399            THE HIGH SCHOOL FOR LANGUAGE AND DIPLOMACY          s   
38  02M427                 MANHATTAN ACADEMY FOR ARTS & LANGUAGE          s   
40  02M437           HUDSON HIGH SCHOOL OF LEARNING TECHNOLOGIES          s   

                    Avg_CR_score Avg Math Score Avg Writing Score  
SNo Database_Number                                                
22  02M392                     s              s                 s  
23  02M393                     s              s                 s  
25  02M399                     s              s                 s  
38  02M427                     s              s                 s  
40  02M437                     s              s                 s

## What data will be used to answer the question? 

- From the above data it can be concluded that there are total 58 schools which either have not submitted SAT Result data or have submitted Junk data due to 's' values in columns.
- Hence the above data can be used to answer the question of Inconsistent values in dataset.

## Describe a second way to answer the same question?

**Tentative reasons for missing data in our dataset**

** 1) School Geographical location:**
- Schools in better regions are more likely to have better resources and facilities for students.
- Therefore schools situated in rural areas did not record the data for SAT score could be a reason for missing data.


** 2) Lack of Participants :**
- Less number of members could likewise be one reason for Lack of information.

- There are 57 such schools which have provided inconsistent data.

- In other words, Few schools might have negligible number of applicants in order to report the SAT results and hence data for such schools are not in an expected format.

## What are assumptions and limitations of the approach?

**Assumptions for Inconsistent data could be:**
- Less Participants: 5 or fewer seniors might have appeared for SAT examination
- Geographical location: Undeveloped regional part or lack of resources


**Limitation of the above Approach:**

- Total 57 number of Schools have not reported SAT result data or have reported inconsistent data which accounts to the 12% of an entire dataset.
- If we select reason for inconsistent data as Number of applicants less than 5 then it would  generate unreliable average for all threee courses.
- Schools which have not reported the data could be having a large number of students.But only few of them must have participated for SAT examination.Thus this accounts for negligible data and hence we get an inconsistent information in dataset.
- Not counting such schools would prove not to be fair for students with highest SAT scores who belongs to such schools. 

## What is the statement that you will be able to make after doing the analysis?

- Scores from all schools must be considered for computing an average of all SAT sections.
- Instead of calculating the average of individual schools, an average of Multiple schools belonging to same regional area could be considered which might compensate for an uncertainity of average result data for that particular school.